app/models/user.rb

```diff
@@ -180,20 +180,18 @@ def send_reset_password_instructions
 
   def update_user_preferences(pref_hash)
     if self.disable_mail
-      UserPreference::VALID_EMAIL_TYPES.each{|x| self.user_preferences.find_or_create_by(email_type: x)}
+      UserPreference::VALID_EMAIL_TYPES.each do |type|
+        user_preferences.find_or_create_by(email_type: type).update(email_enabled: false)
+      end
       self.disable_mail = false
       self.save
     end
 
     pref_hash.keys.each do |key|
-      if pref_hash[key] == 'true'
-        self.user_preferences.find_or_create_by(email_type: key)
-      else
-        block = user_preferences.find_by(email_type: key)
-        if block
-          block.destroy
-        end
-      end
+      email_enabled = pref_hash[key] == "false"
+      user_preferences
+        .find_or_create_by(email_type: key)
+        .update(email_enabled: email_enabled)
```

Comment:
@SuperTux88 quick implementation question, while I haven't gone too far.

So the way I've written it here, once the user submits email configuration we create all possible records for user preferences per notification type. 

So even if I as a user uncheck just one checkbox and save, I'll still have 9 records created, the rest 8 of which match the default value. So the records are absent until the moment when the user first interacts with the email settings and after that they are created and stay in the DB forever.

Alternatively I could make the behavior here a bit more sophisticated and only create the records when the behavior is different from the default one.

In both cases if the user has never interacted with these settings they don't have user preferences records and the absence means the default behavior. But should we intentionally reduce the number of records by not creating the ones that match the default?

I like to keep the code simpler and keep how it works here -- create records for the user even if they match the default. 

But what do you think?

Reply:
Yes, that sounds good to me. I don't think the size of the table should become a problem (and if it does, we can still easily clean it up again in the future by just deleting all the rows that match the default).

Also I like that with the way you have done it now, if the user didn't touch the settings, it stays the default, but as soon as the user changes something, we know that they intentionally wanted to keep the others enabled, so it's fine to store them, even if it's the current default. If we would change the default in the future, then it would only change for the users who didn't touch any settings, but all users who intentionally set their preferences keep it the way they set it.

Reply:
Sounds good,  thanks! There are a few tests that set expectations around number of records so I wanted to clarify that before updating it everywhere.